### PR TITLE
Closes #28 | Add SEA Wiki loader

### LINKED ISSUE #28: Create dataset loader for SEA Wikipedia
Dataloader name: `sea_wiki/sea_wiki.py`
DataCatalogue: https://seacrowd.github.io/seacrowd-catalogue/card.html?SEA%20Wikipedia

|     Dataset| sea_wiki |
|-------------|---|
| Description | SEA Lang & Local Langs Wikipedia Archives, dumped from WIkipedia HF and processed by boilerplate removal. This dataset consists of URL of referred Wikipedia Article, its Title, and its Text Data (Article Contents). |
| Subsets | ace, ban, bcl, bjn, bug, cbk_zam, gor, id, ilo, jv, km, lo, mad, map_bms, min, mnw, ms, nia, pag, pam, shn, su, tet, th, tl, vi, war |
| Languages | ace, ban, bcl, bjn, bug, cbk, gor, ilo, ind, jav, khm, lao, mad, min, mnw, mya, nia, pag, pam, shn, sun, tet, tgl, tha, vie, war |
| License | Creative Commons Attribution Share Alike 4.0 (cc-by-sa-4.0) |
| Homepage | https://huggingface.co/datasets/sabilmakbar/sea_wiki |
| HF URL | https://huggingface.co/datasets/sabilmakbar/sea_wiki |
| Paper URL | No publication |

### PR DESCRIPTION
**Pull request by sabilmakbar**
Closes #28 

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

**Comment by sabilmakbar** [comment]
Another test-cases that went successful of ```load_dataset``` method with various config names. This ```py``` script is placed at base dir of this repo

```
import json
from itertools import product

import datasets
from datasets import load_dataset

def unit_tests_seawiki_load_dset(dset: datasets.Dataset | datasets.DatasetDict):
  if isinstance(dset, datasets.DatasetDict):
    if len(dset.keys()) != 1:
      raise AssertionError("Unexpected #keys!")
    if list(dset.keys()) != ["train"]:
      raise AssertionError("Unexpected keys name!")
  
    dset = dset["train"]

  if dset.num_rows <= 0:
    raise AssertionError("The number of datasets isn't expected!")

if __name__ == "__main__":
  default_config_templates = ["sea_wiki_source", "sea_wiki_seacrowd", "sea_wiki_source_ssp", "sea_wiki_source_t2t", "sea_wiki_seacrowd_ssp", "sea_wiki_seacrowd_t2t"]
  for default_config in default_config_templates:
    print(f"Running `load_dataset` for default config {default_config}")
    dset = load_dataset(
      "/Users/salsabil.akbar/Documents/seacrowd-datahub-sabil-clone/seacrowd/sea_datasets/sea_wiki/sea_wiki.py",
      default_config
    )
    unit_tests_seawiki_load_dset(dset)

  with open("seacrowd/sea_datasets/sea_wiki/lang_config.json", "r") as f:
    langs = list(json.load(f).keys())

  config_templates = ["sea_wiki_source_{lang}_ssp", "sea_wiki_source_{lang}_t2t", "sea_wiki_seacrowd_{lang}_ssp", "sea_wiki_seacrowd_{lang}_t2t"]

  for _config, lang in product(config_templates, langs):
    config = _config.format(lang=lang)
    print(f"Running `load_dataset` for config {config} and lang {lang}")
    load_dataset(
      "/Users/salsabil.akbar/Documents/seacrowd-datahub-sabil-clone/seacrowd/sea_datasets/sea_wiki/sea_wiki.py",
      config
    )
    unit_tests_seawiki_load_dset(dset)
```

**Review by holylovenia — CHANGES_REQUESTED** [review]
Good work, the dataloader works well! One little thing, could you please change the `text_1_name` and `text_2_name` in the `t2t` schemas to `document` and `title`? @sabilmakbar

**Comment by sabilmakbar** [comment]
Thanks for the review, @holylovenia. Made the changes based on ur suggestion

**Review by holylovenia — APPROVED** [review]
Hi @sabilmakbar, thanks for the changes! LGTM. Let's wait for @yongzx's review then we can merge.

**Comment by yongzx** [comment]
Hi @sabilmakbar, thanks for the PR. I've ran the unittests and the dataloader. The implementations work as intended. 

I have a couple questions/comments:
- Is there a documentation of what the config names are for this PR? For instance, what does "sea_wiki_seacrowd" refer to? Is it expected that it has fewer data instances than `"sea_wiki_seacrowd_sun_t2t"` (what does the `sun_t2t` refer to?) How is `sea_wiki_source_*` different from `sea_wiki_seacrowd`?
- I also notice that the current dataloader is used differently than https://huggingface.co/datasets/sabilmakbar/sea_wiki (where the latter supports passing in the keyword argument `lang`). Since the dataloader development for this project doesn't require the support for keyword argument, @holylovenia I suppose this is expected right? I'm asking because I am thinking how users from HF can immediately use our dataloaders without much changes, unless I've missed certain decisions on how to enable HF users to get on board with our dataloaders here.

File: seacrowd/sea_datasets/sea_wiki/sea_wiki.py
Comment (yongzx) [inline seacrowd/sea_datasets/sea_wiki/sea_wiki.py:55]: Can we add comments on what `ssp` and `t2t` refer to?

File: seacrowd/sea_datasets/sea_wiki/sea_wiki.py
Comment (yongzx) [inline seacrowd/sea_datasets/sea_wiki/sea_wiki.py:71]: Can we have a comment string on the difference between `source` and `seacrowd`?

File: seacrowd/sea_datasets/sea_wiki/sea_wiki.py
Comment (yongzx) [inline seacrowd/sea_datasets/sea_wiki/sea_wiki.py:216]: Since `"text_1_name": "document", "text_2_name": "title"` is hardcoded, why can't we write `yield id_, {"id": id_, "document": _data["text"], "title": _data["title"]}` instead? Maybe I am missing something here (potential conflicts with line 208?) or perhaps there's foreseeable extension of this schema where `text_1_name` would refer to other document type?

**Review by yongzx — CHANGES_REQUESTED** [review]

**Comment by sabilmakbar** [comment]
Thanks for the thoughtful review, @yongzx!

Maybe I'll answer all of your inquiries in one go, not for each inquiry, as I find those are quite inter-lapped (please bear with me for a while :D)

This SEACrowd dataloader intends to centralize and index the datasets across SEA NLP datasets, and because it can have various tasks covered and even different datasets with the same task can have different schema, SEACrowd allows the dataloader implementation to have 2 different config types, namely `source` and `seacrowd`. The `source` type allows the intended user of SEACrowd Dataloader to load the data as its original schema, whereas the `seacrowd` schema has to follow the unified schema for each task that this dataset intends for.

Specifically, for this task, since the submission for this Dataloader hints that this dataset can be used for 2 different tasks (Semi-Supervised Pretrained and Summarization/Abstraction, you may find the `if-else` cases in different methods throughout this dataloader (`_info()`, `_generate_examples()`, and `_split_generators()`

Related to the flag of `ssp` and `t2t` refers to, you may check `seacrowd/utils/constants.py` folder and digest its config to understand it wholly. But the short explanation is, that `ssp` and `t2t` refer to the config names suffixes that should be implemented (bcs it's being tested in test-suites after this dataloader script has been written). It refers to SEACrowd config suffixes for Self-Supervised Pretraining Task (SSP Schema) and for Summarization Task (T2T Schema), respectively.

Related to your 2nd qns about passing additional kwargs, you're right. Since this uses `SEACrowdConfig` from `seacrowd/utils/configs.py`, it doesn't support any additional kwargs (yet), since it might add some complexity to it. Hence the loader implementation is a bit different from what you see in the HF Source Repo.

For the last one, since this dataset is quite massive in size if it's being loaded for all available languages, my intention or design for the default schema of `sea_wiki_seacrowd_t2t` is to get the samples and a glimpse of how the data looks like, hence just choosing one local language rather than a national lang or all languages loaded (in SEA, most of the folks doesn't have excellent computing powers, both in Memory and Storage). Fortunately, the first arg in `lang_config.json` is a local lang of `ace` (Acehnese), hence implicitly, running the config of `sea_wiki_seacrowd_ace_t2t` will yield the same result to `sea_wiki_ace_seacrowd_t2t`. Also, if no specific task is being provided in its config name (like running `sea_wiki_source`), it will choose Summarization as the default task (L91) of the code.

Hope it answers all of your questions, and I'll be more than happy to discuss further!

**Review by yongzx — APPROVED** [review]

File: seacrowd/sea_datasets/sea_wiki/sea_wiki.py
Comment (yongzx) [reply seacrowd/sea_datasets/sea_wiki/sea_wiki.py:55]: Addressed.

File: seacrowd/sea_datasets/sea_wiki/sea_wiki.py
Comment (yongzx) [reply seacrowd/sea_datasets/sea_wiki/sea_wiki.py:71]: Addressed.

**Comment by yongzx** [comment]
@sabilmakbar Thanks for the explanations and it helps immensely. I've changed the review status to approved.

**Comment by sabilmakbar** [comment]
Appreciate it, @yongzx!

btw, @holylovenia or @yongzx , if you're going to merge this, would that be okay to choose `squash and merge` to the `master` branch instead of `create a merge commit` to keep the history on `master` branch rather clean?

also, for this PR it shd be you two that do the merge action rather than me bcs both of you are the reviewers 😄

**Comment by sabilmakbar** [comment]
btw do you think we shd add the more detailed comments on several parts of the dataloader that you mentioned to give more clarity, @yongzx ?

Since Holy didn't asked for this due to her familiarity of the core codebase 😆 and I think having you as the reviewer gives a better perspective regarding the clarity on this dataloader script

**Comment by yongzx** [comment]
> one more, do you think we shd add the more detailed comments on several parts of the dataloader that you mentioned to give more clarity, @yongzx ?
> 
> Since Holy didn't asked for this due to her familiarity of the core codebase 😆 and I think having you as the reviewer gives a better perspective regarding the clarity on this dataloader script

In my experience, users wouldn't check the dataloader unless the documentation on the intended/unintended behaviors is unclear. So originally how I tested this (and I assume new users as well) was to run the dataset loader script without any config, and see what config this dataset loader support, and whether I can understand what the config refers to. What I did was to refer to the `task_schemas.md` to understand the abbreviations. I could find `t2t` there but not `ssp`. We probably need an update there. `source` vs `seacrowd` distinction was something I've picked up on after checking your code and also the design of the custom schema for a question I had before: why `text_1` `text_2` instead of `document` `title`?

All in all, I think the current dataloader script is good; hence ignore my comments about ssp and t2t as well as source/seacrowd. We perhaps need a succinct centralized one-stop documentation to point users to docs that easily allows user to make sense of the data. E.g., going to `seacrowd/utils/constants.py` is less user-friendly than having them on `task_schemas.md`. E.g., FAQ to answer the difference between `source` and `seacrowd`. 

However, this isn't the most urgent matter – it's something we need to keep in mind later on the user-centric design of our repo. What do you think @sabilmakbar @holylovenia ?

**Comment by sabilmakbar** [comment]
> In my experience, users wouldn't check the dataloader unless the documentation on the intended/unintended behaviors is unclear. So originally how I tested this (and I assume new users as well) was to run the dataset loader script without any config, and see what config this dataset loader support, and whether I can understand what the config refers to. What I did was to refer to the `task_schemas.md` to understand the abbreviations. I could find `t2t` there but not `ssp`. We probably need an update there. `source` vs `seacrowd` distinction was something I've picked up on after checking your code and also the design of the custom schema for a question I had before: why `text_1` `text_2` instead of `document` `title`?
> 
> All in all, I think the current dataloader script is good; hence ignore my comments about ssp and t2t as well as source/seacrowd. We perhaps need a succinct centralized one-stop documentation to point users to docs that easily allows user to make sense of the data. E.g., going to `seacrowd/utils/constants.py` is less user-friendly than having them on `task_schemas.md`. E.g., FAQ to answer the difference between `source` and `seacrowd`.
> 
> However, this isn't the most urgent matter – it's something we need to keep in mind later on the user-centric design of our repo. What do you think @sabilmakbar @holylovenia ?

I think that's perfectly articulated regarding the ideal user-centric repo, except for the last point regarding the urgency on this. With a lot of submissions coming up (and prob a lot of dataloader implementations), I think this should be the main focus of the maintainers (us), on giving a clear & concise guideline regarding task definitions, allowed/compulsory configs, review checklists, and similar nitty-gritty kinds of stuff that enable us to raise the standards of code and its understanding across the users (not just Dataloader contributors, but also end-users who want to simply load the data from this data catalog).

**Comment by holylovenia** [comment]
@yongzx @sabilmakbar Yes, it's definitely our intention to make the documentation as clear, up-to-date, and user-centric as possible. I think we'll need to clean up (especially the older documentations) and sort out the examples & definitions, the sooner the better. That being said, all available hands are being put on the datasheet & dataloader reviewing for now. Let's put your suggestions on the backlog so we can come back to it and work on it once we can.

**Pull request merged**
